Stratis.Bitcoin/Features/BlockStore/LoopSteps/DownloadBlockStep.cs

```diff
@@ -0,0 +1,176 @@
+﻿using NBitcoin;
+using Stratis.Bitcoin.BlockPulling;
+using Stratis.Bitcoin.BlockStore;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Stratis.Bitcoin.Features.BlockStore.LoopSteps
+{
+    // Continuously download blocks until a stop condition is found.
+    // There are two operations:
+    //      Find blocks to download by asking them from the BlockPuller
+    //      Collecting downloaded blocks and persisting them as a batch to the BlockRepository
+
+    internal sealed class DownloadBlockStep : BlockStoreLoopStep
+    {
+        internal DownloadBlockStep(BlockStoreLoop blockStoreLoop)
+            : base(blockStoreLoop)
+        {
+        }
+
+        private int insertBlockSize = 0;
+        private int stallCount = 0;
+        private List<BlockPair> blockPairsToStore = new List<BlockPair>();
+
+        internal override async Task<BlockStoreLoopStepResult> Execute(ChainedBlock nextChainedBlock, CancellationToken cancellationToken, bool disposeMode)
+        {
+            if (disposeMode)
```

Comment:
I cant refactor this right now as some of the logic in the steps depends on it. I.e not every step will NOT execute if disposeMode is true.

We can possible refactor this in the next PR?